File: src/main/java/com/technews/scheduler/SubscribeMailScheduler.java
package com.technews.scheduler;

import com.technews.aggregate.posts.constant.PostSubjects;
import com.technews.aggregate.posts.domain.Post;
import com.technews.aggregate.posts.domain.repository.PostsRepository;
import com.technews.aggregate.releases.domain.Release;
import com.technews.aggregate.releases.domain.repository.ReleasesRepository;
import com.technews.aggregate.subscribe.dto.SubscriberMailContents;
import com.technews.aggregate.subscribe.service.SubscribeService;
import com.technews.common.config.event.Events;
import com.technews.common.constant.JdkVersion;
import com.technews.common.constant.SpringRepository;
import com.technews.common.event.dto.SendMailEvent;
import com.technews.common.util.MailTemplateUtils;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Slf4j
@Component
@RequiredArgsConstructor
public class SubscribeMailScheduler {

    private final static String SUBJECT = "Java & Spring News";

    private final ReleasesRepository releasesRepository;
    private final PostsRepository postsRepository;
    private final SubscribeService subscribeService;

    @Scheduled(cron = "0 0 9 * * *")
    public void runScheduler() {
        final List<Release> releases = releasesRepository.findBySharedFalse();
        final List<Post> posts = postsRepository.findBySharedFalse();
        SubscriberMailContents subscriberMailContents = getSubscriberMailContents(releases, posts);

        final List<String> subscriber = subscribeService.findSubscriberEmail();
        Events.raise(SendMailEvent.builder()
                .subject(SUBJECT)
                .contents(MailTemplateUtils.generateContents())
                .addressList(Optional.of(subscriber))
                .build());
    }

    private SubscriberMailContents getSubscriberMailContents(final List<Release> releases, final List<Post> posts) {
        List<Release> releaseOfSpring = getReleaseOfSpring(releases);
        List<Release> releaseOfJava = getReleaseOfJava(releases);
        List<Post> postOfSpring = getPostOfSpring(posts);
        List<Post> postOfJava = getPostOfJava(posts);

        return SubscriberMailContents.builder()
                .releaseOfSpring(releaseOfSpring)
                .releaseOfJava(releaseOfJava)
                .postOfSpring(postOfSpring)
                .postOfJava(postOfJava)
                .build();
    }

    private List<Release> getReleaseOfSpring(final List<Release> releases) {
        return releases.stream()
                .filter(release -> SpringRepository.list().contains(release.getProject()))
                .collect(Collectors.toList());
    }

    private List<Release> getReleaseOfJava(final List<Release> releases) {
        return releases.stream()
                .filter(release -> JdkVersion.list().contains(release.getProject()))
                .collect(Collectors.toList());
    }

    private List<Post> getPostOfSpring(final List<Post> posts) {
        return posts.stream()
                .filter(post -> PostSubjects.SPRING.value().equals(post.getSubject()))
                .collect(Collectors.toList());
    }

    private List<Post> getPostOfJava(final List<Post> posts) {
        return posts.stream()
                .filter(post -> PostSubjects.JAVA.value().equals(post.getSubject()))
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/technews/scheduler/SubscribeMailScheduler.kt
package com.technews.scheduler

import com.technews.aggregate.posts.constant.PostSubjects
import com.technews.aggregate.posts.domain.Post
import com.technews.aggregate.posts.domain.repository.PostsRepository
import com.technews.aggregate.releases.domain.Release
import com.technews.aggregate.releases.domain.repository.ReleasesRepository
import com.technews.aggregate.subscribe.dto.SubscriberMailContents
import com.technews.aggregate.subscribe.service.SubscribeService
import com.technews.common.config.event.Events
import com.technews.common.constant.JdkVersion
import com.technews.common.constant.SpringRepository
import com.technews.common.event.dto.SendMailEvent
import mu.KotlinLogging
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

private val logger = KotlinLogging.logger {}

@Component
class SubscribeMailScheduler(
    private val releasesRepository: ReleasesRepository,
    private val postsRepository: PostsRepository,
    private val subscribeService: SubscribeService,
) {
    @Scheduled(cron = "0 0 9 * * *")
    fun runSchedule() {
        logger.info("[SubscribeMailScheduler] Start...")
        val releases = releasesRepository.findBySharedFalse()
        val posts = postsRepository.findBySharedFalse()
        if (releases.isEmpty() && posts.isEmpty()) {
            logger.info("[SubscribeMailScheduler] There are no news")
            return
        }

        val subscriberMailContents = createSubscriberMailContents(releases, posts)
        val subscriber = subscribeService.findSubscriberEmail()
        Events.raise(
            SendMailEvent(
                subject = SUBJECT,
                contents = subscriberMailContents,
                addressList = subscriber,
            ),
        )

        updateShared(releases, posts)
    }

    private fun createSubscriberMailContents(
        releases: List<Release>,
        posts: List<Post>,
    ): SubscriberMailContents {
        val releaseOfSpring = getReleaseOfSpring(releases)
        val releaseOfJava = getReleaseOfJava(releases)
        val postOfSpring = getPostOfSpring(posts)
        val postOfJava = getPostOfJava(posts)

        return SubscriberMailContents(
            springReleases = releaseOfSpring,
            javaReleases = releaseOfJava,
            springPosts = postOfSpring,
            javaPosts = postOfJava,
        )
    }

    private fun getReleaseOfSpring(releases: List<Release>): List<Release> =
        releases.filter { SpringRepository.list().contains(it.project) }

    private fun getReleaseOfJava(releases: List<Release>): List<Release> =
        releases.filter { JdkVersion.list().contains(it.project) }

    private fun getPostOfSpring(posts: List<Post>): List<Post> =
        posts.filter { PostSubjects.SPRING.value == it.subject }

    private fun getPostOfJava(posts: List<Post>): List<Post> =
        posts.filter { PostSubjects.JAVA.value == it.subject }

    private fun updateShared(
        releases: List<Release>,
        posts: List<Post>,
    ) {
        releases.forEach { it.updateShare() }
        posts.forEach { it.updateShare() }
    }

    private fun Release.updateShare() {
        this.share()
        releasesRepository.save(this)
    }

    private fun Post.updateShare() {
        this.share()
        postsRepository.save(this)
    }

    companion object {
        private const val SUBJECT = "Java & Spring News"
    }
}
